### ISSUE TITLE
Error 500 on Callback stage of OIDC-based SSO flow on Azure

### ISSUE THREAD

**Issue by steevaavoo** [CloudFire-LLC/cloudfire#1684]
**Description**
Firezone server returns `Error 500: unable to handle this request` after authentication to Azure has been completed and the Callback URI is used.

**To Reproduce**

1. Configure an Azure OIDC Provider per the guide [here](https://www.firezone.dev/docs/authenticate/oidc/azuread)
2. Navigate to Firezone Web UI and click "Azure SSO" (or whatever you've named it)
3. Enter Azure AD credentials to authenticate with AAD
4. Click `Sign in`
5. Error page occurs

**Expected behavior**
Rather than the 500 error, I expect to be presented with the Firezone Web UI with my chosen user logged in.

**Estimated Impact**
High: This prevents us from viably using the product.

**Screenshots / Logs**
The below was grabbed from the Firezone Service Container log during such an attempt:

```
firezone-firezone-1  | ** (exit) an exception was raised:
firezone-firezone-1  |     ** (Protocol.UndefinedError) protocol Phoenix.HTML.Safe not implemented for %FunctionClauseError{module: FzHttpWeb.UserFromAuth, function: :find_or_create, arity: 2, 
kind: nil, args: nil, clauses: nil} of type FunctionClauseError (a struct). This protocol is implemented for the following type(s): Atom, BitString, Date, DateTime, Decimal, Float, FzHttp.Types.IPPort, Integer, List, NaiveDateTime, Phoenix.HTML.Form, Phoenix.LiveComponent.CID, Phoenix.LiveView.Component, Phoenix.LiveView.Comprehension, Phoenix.LiveView.JS, Phoenix.LiveView.Rendered, 
Postgrex.INET, Time, Tuple, URI
firezone-firezone-1  |         (phoenix_html 3.3.1) lib/phoenix_html/safe.ex:1: Phoenix.HTML.Safe.impl_for!/1
firezone-firezone-1  |         (phoenix_html 3.3.1) lib/phoenix_html/safe.ex:15: Phoenix.HTML.Safe.to_iodata/1
firezone-firezone-1  |         (phoenix 1.7.2) lib/phoenix/controller.ex:1010: anonymous fn/5 in Phoenix.Controller.template_render_to_iodata/4
firezone-firezone-1  |         (telemetry 1.2.1) /app/deps/telemetry/src/telemetry.erl:321: :telemetry.span/3
firezone-firezone-1  |         (phoenix 1.7.2) lib/phoenix/controller.ex:976: Phoenix.Controller.render_and_send/4
firezone-firezone-1  |         (phoenix 1.7.2) lib/phoenix/endpoint/render_errors.ex:84: Phoenix.Endpoint.RenderErrors.instrument_render_and_send/5
firezone-firezone-1  |         (phoenix 1.7.2) lib/phoenix/endpoint/render_errors.ex:62: Phoenix.Endpoint.RenderErrors.__catch__/5
firezone-firezone-1  |         (plug_cowboy 2.6.1) lib/plug/cowboy/handler.ex:11: Plug.Cowboy.Handler.init/2
```

**Platform (please complete the following information):**
 - Firezone Version `0.7.30` - confirmed also occurring in `0.7.31`
 - OS: `Linux`
 - Distro: `Ubuntu 22.04`
 - Kernel version: `5.15.0-1040-azure`
 - Deployment method: `Docker`

**Comment by jamilbk** [comment]
Thanks for filing the report @steevaavoo. #1686 will likely fix this.

In the meantime, are you sure the `email` claim is being passed in your ID token? Firezone needs that to match the user.

**Comment by steevaavoo** [comment]
Hey Jamil. Thanks for the reply and suggestion.

So in this case, the users don't strictly _have_ "Primary email addresses" in the Exchange mailbox sense. They have their UPN which is in the traditional email address format.

Could it be that the Microsoft Graph `email` property has nothing to return in this use case, so that mangles the token? Would it be useful (feasible?) to change it to `userPrincipalName` instead? Might lend some resilience for these cases.

**Comment by jamilbk** [comment]
@steevaavoo Unfortunately Microsoft doesn't play nicely here. `UPN` is only used for AD.

Every other provider returns an `email` in the claims when the `email` scope is requested. `email` is the only universal attribute among OIDC providers, so that's what we use.

In this case it's not that the token is mangled, it's that Azure is simply not providing the `email` claim. Rather than build provider-specific codepaths for issues like this, we like to suggest for you to adapt your OIDC connector on the Azure side to pass the expected `email` claim.

This doc may be helpful if you haven't seen it yet: https://www.firezone.dev/docs/authenticate/oidc/azuread

**Comment by steevaavoo** [comment]
I was just coming back to say "oh dear, Microsoft doesn't _have_ UPN in Graph" and saw the above. Thanks for that!

**Comment by steevaavoo** [comment]
Nailed it! 

In case it helps others in similar circumstances: We copied the user principal name in the AAD portal to be the Email Address property of the user in question - and the login succeeded immediately.

That error message you're adding in will hopefully lead people to this same conclusion. 

I'm happy to put together a quick guide on what we did as a potential solution for others, if that would be useful?

Great work Jamil, and thanks again!
